mm2src/coins/z_coin/storage/walletdb/mod.rs

```diff
@@ -33,27 +48,74 @@ pub struct WalletDbShared {
     ticker: String,
 }
 
+#[cfg(not(target_arch = "wasm32"))]
+async fn checkpoint_block_from_height(
+    height: u64,
+    rpc: &mut impl ZRpcOps,
+) -> MmResult<Option<CheckPointBlockInfo>, WalletDbError> {
+    let tree_state = rpc
+        .get_tree_state(height)
+        .await
+        .mm_err(|err| WalletDbError::GrpcError(err.to_string()))?;
+    let hash = H256::from_str(&tree_state.hash)
+        .map_err(|err| WalletDbError::DecodeError(err.to_string()))?
+        .reversed();
+    let sapling_tree = Bytes::new(
+        FromHex::from_hex(&tree_state.tree).map_err(|err: FromHexError| WalletDbError::DecodeError(err.to_string()))?,
+    );
+
+    Ok(Some(CheckPointBlockInfo {
+        height: tree_state.height as u32,
+        hash,
+        time: tree_state.time,
+        sapling_tree,
+    }))
+}
+
 #[cfg(not(target_arch = "wasm32"))]
 impl<'a> WalletDbShared {
-    pub async fn new(zcoin_builder: &ZCoinBuilder<'a>) -> MmResult<Self, WalletDbError> {
+    pub async fn new(zcoin_builder: &ZCoinBuilder<'a>, rpc: &mut impl ZRpcOps) -> MmResult<Self, WalletDbError> {
         let z_spending_key = match zcoin_builder.z_spending_key {
             Some(ref z_spending_key) => z_spending_key.clone(),
             None => extended_spending_key_from_protocol_info_and_policy(
                 &zcoin_builder.protocol_info,
                 &zcoin_builder.priv_key_policy,
             )
-            .map_err(|err| WalletDbError::ZCoinBuildError(err.to_string()))?,
+            .mm_err(|err| WalletDbError::ZCoinBuildError(err.to_string()))?,
+        };
+
+        let current_block_height = rpc
+            .get_block_height()
+            .await
+            .map_err(|err| WalletDbError::GrpcError(err.to_string()))?;
+        let sync_block = match &zcoin_builder.z_coin_params.mode {
+            ZcoinRpcMode::Light { sync_params, .. } => {
+                let sync_height = match sync_params {
+                    Some(SyncStartPoint::Date(date)) => zcoin_builder
+                        .calculate_starting_height_from_date(*date, current_block_height)
+                        .mm_err(WalletDbError::UtxoCoinBuildError)?,
+                    Some(SyncStartPoint::Height(height)) => *height,
+                    None => zcoin_builder
+                        .calculate_starting_height_from_date(now_sec() - DAY_IN_SECONDS, current_block_height)
+                        .mm_err(WalletDbError::UtxoCoinBuildError)?,
+                };
+
+                checkpoint_block_from_height(sync_height, rpc).await?
+            },
+
+            ZcoinRpcMode::Native => zcoin_builder.protocol_info.check_point_block.clone(),
```

Comment:
This entire code block should likely be in another place. `WalletDb::new` is now strictly dependent on RPC client, the error variants are mixed up like `WalletDbError` having `GrpcError` and so on. In general, it's not `WalletDb` responsibility to call any of the RPC methods. Instead, I advise replacing `rpc` arg with just `checkpoint_block`.

Reply:
noted. thanks